### ISSUE TITLE
Can't get refresh to trigger

### ISSUE THREAD

**Issue by drmmr763** [Flyrell/axios-auth-refresh#21]
Hi There

First off, I just want to say a huge thanks for this utility. It's exactly what I'm looking for. Unfortunately, I just can't seem to make it work quite right. I'm sure I'm close and missing something obvious. 

*Context*
* My app allows a user to log in and is granted a short-lived access token and refresh token
* I'm using the `axios.create` method to instantiate the baseURL of the api
* I have two custom interceptors applied to the axios instance that automatically sign the request with the access token

*Observed Behavior*
* Once the short-lived access token expires, any additional requests that require authentication (except one, more below) following the expiration will correctly return a 401 error response. 
* While inspecting the network tab, the refresh doesn't seem to be triggering. 

*Debug Steps*
* Removed my custom interceptors from the equation
* Removed the axios instantiation with `axios.create` and just used Axios's default export.
* added `.then, .catch and .finally` to the api action to try force promises to go through. 

I'm out of ideas of things to try at this point. My other interceptors are working great. I'm sure there's something minor I'm doing wrong with promises or something like that. Any tip that could be provided to point me in the right direction would be greatly appreciated! 

Chad

Code:
```
api.json

import axios from "axios";
import store from "../store";
import { userActions } from "../_actions/user.actions";
import createAuthRefreshInterceptor from "axios-auth-refresh";
import { showLoading, hideLoading } from "react-redux-loading-bar";

const { dispatch } = store;

const api = axios.create({
    baseURL: process.env.REACT_APP_API_ROOT,
    timeout: 150000,
    headers: {
        Accept: "application/json"
    }
});

// Function that will be called to refresh authorization
const refreshAuthLogic = (failedRequest) => {
    console.log('failed request interceptor');

    // unable to refresh because the token is not available
    if (userActions.getRefreshToken() === false) {
        return Promise.reject(failedRequest)
    }

    api
        .post("/oauth/token", {
            grant_type: "refresh_token",
            refresh_token: userActions.getRefreshToken(),
            client_id: process.env.REACT_APP_CLIENT_ID,
            client_secret: process.env.REACT_APP_CLIENT_SECRET
        })
        .then(tokenRefreshResponse => {
            // save the new access and refresh tokens
            localStorage.setItem(
                "auth-token",
                JSON.stringify(tokenRefreshResponse.data)
            );
            failedRequest.response.config.headers["Authentication"] = "Bearer " + userActions.getAuthenticatedToken();
            return Promise.resolve();
        })
}


// Instantiate the interceptor (you can chain it as it returns the axios instance)
createAuthRefreshInterceptor(api, refreshAuthLogic, {statusCodes: [ 401, 400 ]});

// show loader
api.interceptors.request.use(function(config) {
    dispatch(showLoading());

    return config;
});

// hide loader
api.interceptors.response.use(
    response => {
        dispatch(hideLoading());
        return response;
    },
    error => {
        dispatch(hideLoading());
        return Promise.reject(error);
    }
);

// append auth header
api.interceptors.request.use(function(config) {
    let token = userActions.getAuthenticatedToken();

    if (token !== false) {
        config.headers.Authorization = `Bearer ${token}`;
    }

    return config;
});

export default api;
```

Actions.js

```
/**
 * Retrieve collection of lead sheets
 * @param page
 * @param orderColumn
 * @param orderDirection
 * @returns {Function}
 */
export function getLeadSheets(page, orderColumn, orderDirection) {
    return dispatch => {
        dispatch(request());

        leadsApi.getLeadSheets(page, orderColumn, orderDirection).then(
            response => {
                dispatch(success(response));
            },
            error => {
                dispatch(failure(error));
            }
        )
    };

    function request() {
        return {
            type: leadsConstants.LEAD_SHEET_INDEX_REQUEST
        };
    }

    function success(response) {
        return {
            type: leadsConstants.LEAD_SHEET_INDEX_SUCCESS,
            payload: response.data
        };
    }

    function failure(error) {
        return {
            type: leadsConstants.LEAD_SHEET_INDEX_FAILURE,
            payload: error
        };
    }
}
```

**Comment by Flyrell** [comment]
Hi Chad,

I'm sorry but I have completely missed your issue. 

I didn't even read the whole thing, so correct me if I'm wrong, but it seems like you're missing the return statement in `refreshAuthLogic` function. 

Try returning the axios call, so the interceptor can bind to it.

```javascript
const refreshAuthLogic = (failedRequest) => {
    console.log('failed request interceptor');

    // unable to refresh because the token is not available
    if (userActions.getRefreshToken() === false) {
        return Promise.reject(failedRequest)
    }

    return api
        .post("/oauth/token", {
            grant_type: "refresh_token",
            refresh_token: userActions.getRefreshToken(),
            client_id: process.env.REACT_APP_CLIENT_ID,
            client_secret: process.env.REACT_APP_CLIENT_SECRET
        })
        .then(tokenRefreshResponse => {
            // save the new access and refresh tokens
            localStorage.setItem(
                "auth-token",
                JSON.stringify(tokenRefreshResponse.data)
            );
            failedRequest.response.config.headers["Authentication"] = "Bearer " + userActions.getAuthenticatedToken();
            return Promise.resolve();
        })
}
```

**Comment by drmmr763** [comment]
Hi Flyrell

Thanks for getting back to me. I did figure this out just the other day. You are right, I was missing the return. 

The other big problem I was having was this: I had a Laravel backend with CORS configured on certain routes. But the auth route did not have CORS configured correctly. This meant whenever the XHR request failed it failed as a network / browser error not as a catchable error inside the Refresh function. 

This helped me work it out:
https://github.com/axios/axios/issues/569#issuecomment-364817273

Thanks for taking the time to respond and spot the error. This is working for me now and I'm really happy I got it working. It feels like a very clean implementation.
